concurrent_test.go

```diff
@@ -0,0 +1,325 @@
+
+	t *testing.T
+}
+
+func (reader *readWorker) run() error {
+	for {
+		select {
+		case <-reader.stopCh:
+			reader.t.Log("Reading transaction finished.")
+			return nil
+		default:
+		}
+
+		err := reader.db.View(func(tx *bolt.Tx) error {
+			b := tx.Bucket(reader.bucket)
+
+			selectedKey := reader.keys[mrand.Intn(len(reader.keys))]
+			initialVal := b.Get([]byte(selectedKey))
+			time.Sleep(randomDurationInRange(reader.minReadInterval, reader.maxReadInterval))
+			val := b.Get([]byte(selectedKey))
+
+			if !reflect.DeepEqual(initialVal, val) {
+				return fmt.Errorf("read different values for the same key (%q), value1: %q, value2: %q",
+					selectedKey, formatBytes(initialVal), formatBytes(val))
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			readErr := fmt.Errorf("[reader error]: %w", err)
+			reader.t.Log(readErr)
+			reader.errCh <- readErr
+			return readErr
+		}
+	}
+}
+
+type writeWorker struct {
+	db *btesting.DB
+
+	bucket []byte
+	keys   []string
+
+	minWriteBytes    int
+	maxWriteBytes    int
+	minWriteInterval time.Duration
+	maxWriteInterval time.Duration
+	errCh            chan error
+	stopCh           chan struct{}
+
+	t *testing.T
+}
+
+func (writer *writeWorker) run() error {
+	for {
+		select {
+		case <-writer.stopCh:
+			writer.t.Log("Writing transaction finished.")
+			return nil
+		default:
+		}
+
+		err := writer.db.Update(func(tx *bolt.Tx) error {
+			b := tx.Bucket(writer.bucket)
+
+			selectedKey := writer.keys[mrand.Intn(len(writer.keys))]
+
+			valueBytes := randomIntInRange(writer.minWriteBytes, writer.maxWriteBytes)
+			v := make([]byte, valueBytes)
+			if _, cErr := crand.Read(v); cErr != nil {
+				return cErr
+			}
+
+			return b.Put([]byte(selectedKey), v)
+		})
+
+		if err != nil {
+			writeErr := fmt.Errorf("[writer error]: %w", err)
+			writer.t.Log(writeErr)
+			writer.errCh <- writeErr
+			return writeErr
+		}
+
+		time.Sleep(randomDurationInRange(writer.minWriteInterval, writer.maxWriteInterval))
+	}
+}
+
+func randomDurationInRange(min, max time.Duration) time.Duration {
+	d := int64(max) - int64(min)
+	d = int64(mrand.Intn(int(d))) + int64(min)
+	return time.Duration(d)
+}
+
+func randomIntInRange(min, max int) int {
+	return mrand.Intn(max-min) + min
+}
+
+func formatBytes(val []byte) string {
+	if utf8.ValidString(string(val)) {
+		return string(val)
+	}
+
+	return hex.EncodeToString(val)
+}
+
+func saveDataIfFailed(t *testing.T, db *btesting.DB) {
+	if t.Failed() {
+		if err := db.Close(); err != nil {
+			t.Errorf("Failed to close db: %v", err)
+		}
+		backupPath := testResultsDirectory(t)
+		targetFile := filepath.Join(backupPath, "db.bak")
+
+		t.Logf("Saving the DB file to %s", targetFile)
+		err := common.CopyFile(db.Path(), targetFile)
+		require.NoError(t, err)
+		t.Logf("DB file saved to %s", targetFile)
+	}
+}
+
+func testResultsDirectory(t *testing.T) string {
+	resultsDirectory, ok := os.LookupEnv("RESULTS_DIR")
+	if !ok {
+		resultsDirectory = "/tmp/"
```

Comment:
`os.MkdirTemp` ? https://pkg.go.dev/os#MkdirTemp

Reply:
I just copied the code from etcd :)

Reply:
MkdirTemp creates a unique directory name by assign a random suffix to name. For robustness tests I preferred to have predictable file locations to be able to provide stable links in documentation and be able to quickly refresh browser set on same path. Downside is that results are overridden by run.

I think it's up to you to pick

Reply:
Thanks both @ptabor and  @serathius . 

> Downside is that results are overridden by run

Updated. Let's use `os.MkdirTemp` to avoid this.